### PR TITLE
feat(solid-query): Solid Query Adapter for TanStack Query

### PR DESCRIPTION
**Pull request by lukesmurray**
This pull request aims to add the TanStack Query adapter for Solid JS. This pull request will add the following key primitives to the package.

-   `createQuery`
-   `createQueries`
-   `createInfiniteQueries`
-   `createMutation`
-   `useIsFetching`
-   `useIsMutating`
-   `useQueryClient`
-   `QueryClient`
-   `QueryClientProvider`

The adapter docs have been updated as well as examples have been added.

## Acknowledgments

The work on solid-query would not have been possible without

- @ardeora
- @lukesmurray
- @jennyckaplan

This is a follow-up to #4195 but rebased on `main` and with commits from `beta` removed per discussion with @TkDodo and @ardeora.

### CODE REVIEW

File: babel.config.js
Comment (TkDodo) [inline babel.config.js:43]: I'm no babel expert, but I think it should be possible to also have one babel config per package? So we would only keep the shared things in top level and then move the framework specific things further down ?

File: rollup.config.ts
Comment (TkDodo) [inline rollup.config.ts:158]: @DamianOsipiuk is this fine?

**Review by TkDodo — COMMENTED** [review]
just noting down some improvements to the whole structure that we could also do in a follow-up. Maybe this gets easier with a task runner like turborepo where we also have an open PR

**Review by TkDodo — CHANGES_REQUESTED** [review]
🚨 one thing we need to do to make the publish script work is to add solid-query to the package list here (at the end I think):

https://github.com/TanStack/query/blob/357ec041a6fcc4a550f3df02c12ecc7bcdefbc05/scripts/config.ts#L5

**Comment by lukesmurray** [comment]
Before this is merged @ardeora has some follow up work that is not included. So please hold off until we have it.

File: packages/solid-query/package.json
Comment (TkDodo) [inline packages/solid-query/package.json:4]: hmm not sure where 4.3.9 is coming from, as the latest release is 4.3.8 ? I think this makes the codesandbox preview build fail!

File: packages/solid-query/package.json
Comment (TkDodo) [inline packages/solid-query/package.json:31]: is this just copy-paste, because I don't think we have codemods yet for solid-query 😅

**Review by TkDodo — COMMENTED** [review]
let's also add one example to `.codesandbox/ci.json` so that we can see a preview build of solid-query on each PR:

https://github.com/TanStack/query/blob/357ec041a6fcc4a550f3df02c12ecc7bcdefbc05/.codesandbox/ci.json#L3

File: packages/solid-query/package.json
Comment (lukesmurray) [reply packages/solid-query/package.json:43]: yes this was copy paste. I started by copying the entire react-query folder so this should be removed.

File: packages/solid-query/package.json
Comment (lukesmurray) [reply packages/solid-query/package.json]: This was my fault, I thought all versions were pinned together, so I took the version from react-query and got a bit too excited, and submitted the PR with bugs.

**Comment by ardeora** [comment]
> 🚨 one thing we need to do to make the publish script work is to add solid-query to the package list here (at the end I think):
> 
> https://github.com/TanStack/query/blob/357ec041a6fcc4a550f3df02c12ecc7bcdefbc05/scripts/config.ts#L5

Thank you! I added solid-query to the packages in the list below

File: packages/solid-query/package.json
Comment (ardeora) [reply packages/solid-query/package.json]: This should have been fixed in the latest changes

File: packages/solid-query/package.json
Comment (ardeora) [reply packages/solid-query/package.json]: I have removed all the codemod bits so we should be good here too

**Comment by ardeora** [comment]
> let's also add one example to `.codesandbox/ci.json` so that we can see a preview build of solid-query on each PR:
> 
> https://github.com/TanStack/query/blob/357ec041a6fcc4a550f3df02c12ecc7bcdefbc05/.codesandbox/ci.json#L3

Yep! I added the typescript example and it's working great here https://codesandbox.io/s/tanstack-query-example-solid-basic-typescript-lhsczt

**Comment by ardeora** [comment]
Alright I think we should be good to go here. All the tests seem to be passing. And I applied the batching state updates which works great now. Thanks for all the feedback here 😄

**Comment by TkDodo** [comment]
exciting 🚀

**Review by TkDodo — APPROVED** [review]

**Comment by aulneau** [comment]
just wanted to pop in and say great work, this is so exciting!

**Comment by lukesmurray** [comment]
Ok everything seems to be ready to go. @TkDodo feel free to merge. Please add @ardeora, @lukesmurray, @jennyckaplan, and @oscartbeaumont as co-authors on the final merge if possible. This was a team effort, led by @ardeora and I would love to see everyone get recognition for their contributions here.

Cheers 🍻

File: packages/solid-query/src/__tests__/createQuery.test.tsx
Comment (TkDodo) [inline packages/solid-query/src/__tests__/createQuery.test.tsx:2683]: the whole expect was removed here: https://github.com/TanStack/query/pull/4211/commits/0224d6b29073d97426503312723af4c2cc50d9ec

I think we should just consistently expect a render count of 3. with no check at all, the test is rather useless:

```suggestion
    await waitFor(() => screen.getByText('count: 2'))

    // Should be 3 instead of 5
    expect(renders).toBe(3)
```

**Comment by TkDodo** [comment]
I was just about to merge this when I took another look at the codesandbox preview build, and it sadly doesn't render anymore:

https://codesandbox.io/s/tanstack-query-example-solid-basic-typescript-17fbsy

The error is:

```
Uncaught ReferenceError: React is not defined
    at QueryClientProvider (QueryClientProvider.jsx:37:5)
    at dev.js:519:12
    at untrack (dev.js:425:12)
    at Object.fn (dev.js:515:37)
    at runComputation (dev.js:695:22)
    at updateComputation (dev.js:680:3)
    at devComponent (dev.js:526:3)
    at createComponent (dev.js:1228:10)
    at App (index.tsx:142:3)
    at dev.js:519:12
```

I was working yesterday for sure :/

**Comment by OrJDev** [comment]
> I was just about to merge this when I took another look at the codesandbox preview build, and it sadly doesn't render anymore:
> 
> https://codesandbox.io/s/tanstack-query-example-solid-basic-typescript-17fbsy
> 
> The error is:
> 
> ```
> Uncaught ReferenceError: React is not defined
>     at QueryClientProvider (QueryClientProvider.jsx:37:5)
>     at dev.js:519:12
>     at untrack (dev.js:425:12)
>     at Object.fn (dev.js:515:37)
>     at runComputation (dev.js:695:22)
>     at updateComputation (dev.js:680:3)
>     at devComponent (dev.js:526:3)
>     at createComponent (dev.js:1228:10)
>     at App (index.tsx:142:3)
>     at dev.js:519:12
> ```
> 
> I was working yesterday for sure :/

same things here,
 it does work tho when using `@adeora/solid-query` instead of `@tanstack/solid-query`

**Comment by TkDodo** [comment]
Yes, we are seeing this, too. This seems to be an upstream problem with `vite-plugin-solid`.

It worked with `2.3.6` but broke with `2.3.7` which was released today.

we actually pinned the plugin version because of this in our official example, which is working:

https://codesandbox.io/s/github/tanstack/query/tree/main/examples/solid/simple?file=/package.json:411-443

**Comment by TkDodo** [comment]
upstream issue:
- https://github.com/solidjs/vite-plugin-solid/issues/57

**Comment by ardeora** [comment]
> > I was just about to merge this when I took another look at the codesandbox preview build, and it sadly doesn't render anymore:
> > https://codesandbox.io/s/tanstack-query-example-solid-basic-typescript-17fbsy
> > The error is:
> > ```
> > Uncaught ReferenceError: React is not defined
> >     at QueryClientProvider (QueryClientProvider.jsx:37:5)
> >     at dev.js:519:12
> >     at untrack (dev.js:425:12)
> >     at Object.fn (dev.js:515:37)
> >     at runComputation (dev.js:695:22)
> >     at updateComputation (dev.js:680:3)
> >     at devComponent (dev.js:526:3)
> >     at createComponent (dev.js:1228:10)
> >     at App (index.tsx:142:3)
> >     at dev.js:519:12
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I was working yesterday for sure :/
> 
> same things here, it does work tho when using `@adeora/solid-query` instead of `@tanstack/solid-query`


Yeah this was an issue with vite-plugin-solid that we just recently resolved. Can you upgrade vite-plugin-solid to v2.3.8? 

https://stackblitz.com/edit/vitejs-vite-csdjj1?file=package.json,src%2Fmain.tsx,src%2FApp.tsx

Here is a working stackblitz sandbox for reference

**Comment by OrJDev** [comment]
> > > I was just about to merge this when I took another look at the codesandbox preview build, and it sadly doesn't render anymore:
> > > https://codesandbox.io/s/tanstack-query-example-solid-basic-typescript-17fbsy
> > > The error is:
> > > ```
> > > Uncaught ReferenceError: React is not defined
> > >     at QueryClientProvider (QueryClientProvider.jsx:37:5)
> > >     at dev.js:519:12
> > >     at untrack (dev.js:425:12)
> > >     at Object.fn (dev.js:515:37)
> > >     at runComputation (dev.js:695:22)
> > >     at updateComputation (dev.js:680:3)
> > >     at devComponent (dev.js:526:3)
> > >     at createComponent (dev.js:1228:10)
> > >     at App (index.tsx:142:3)
> > >     at dev.js:519:12
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I was working yesterday for sure :/
> > 
> > 
> > same things here, it does work tho when using `@adeora/solid-query` instead of `@tanstack/solid-query`
> 
> Yeah this was an issue with vite-plugin-solid that we just recently resolved. Can you upgrade vite-plugin-solid to v2.3.8?
> 
> https://stackblitz.com/edit/vitejs-vite-csdjj1?file=package.json,src%2Fmain.tsx,src%2FApp.tsx
> 
> Here is a working stackblitz sandbox for reference

Yes it is ok i managed to get it working and added it as an addon to [Create JD App](https://github.com/OrJDev/create-jd-app),

great work you did there.

**Comment by ardeora** [comment]
@all-contributors
please add @lukesmurray  for code. 
please add @jennyckaplan  for code.
please add @oscartbeaumont for code.

**Comment by ardeora** [comment]
@all-contributors
please add @jennyckaplan  for code.

**Comment by ardeora** [comment]
@all-contributors
please add @oscartbeaumont for code.

**Comment by vyktoremario** [comment]
Why does this break for every patch version of `vite-plugin-solid`?
It seems as if `2.3.9` which is the latest version is causing the same error as above.

**Comment by lukesmurray** [comment]
I'm not sure, and I'm sorry you're having that issue. I suggest posting an issue to `vite-plugin-solid` with some information about the error you're getting. If you're getting an error, it's likely other packages are getting errors as well, and my guess is it is an upstream issue rather than a problem with solid-query. But if you suspect it is an issue with solid-query feel free to open an issue with some more information about the error in this repo, and we will take a look.

**Comment by milahu** [comment]
why is it called `createQuery` not `useQuery` like in `@tanstack/query`?

should be the same API no?

**Comment by milahu** [comment]
`@tanstack/solid-query-persist-client` draft

because caching + persistence = profit

<details>

```jsx
/*

@tanstack/solid-query-persist-client



usage

this is a drop-in replacement for QueryClientProvider from @tanstack/solid-query

function App() {
  return (
    <PersistQueryClientProvider
        client={queryClient}
        persister={createPersister()}
      >
      <Main/>
    </PersistQueryClientProvider>
  )
}



see also

https://tanstack.com/query/v4/docs/plugins/persistQueryClient#persistqueryclientprovider



based on

QueryClientProvider in
node_modules/@tanstack/solid-query/build/lib/QueryClientProvider.esm.js
node_modules/@tanstack/solid-query/build/solid/QueryClientProvider.jsx

PersistQueryClientProvider in
node_modules/@tanstack/react-query-persist-client/build/lib/PersistQueryClientProvider.esm.js

persistQueryClientSubscribe in
node_modules/@tanstack/query-persist-client-core/build/lib/persist.esm.js

*/



import { createContext, mergeProps, onMount, onCleanup } from 'solid-js';

import { persistQueryClientSubscribe } from "@tanstack/query-persist-client-core"

import { defaultContext } from "@tanstack/solid-query"



// TODO?
// export { persistQueryClient, persistQueryClientRestore, persistQueryClientSave, persistQueryClientSubscribe }
// export { removeOldestQuery }
//export * from '@tanstack/query-persist-client-core';



const QueryClientSharingContext = createContext(false); // If we are given a context, we will use it.
// Otherwise, if contextSharing is on, we share the first and at least one
// instance of the context across the window
// to ensure that if Solid Query is used across
// different bundles or microfrontends they will
// all use the same **instance** of context, regardless
// of module scoping.



function getQueryClientContext(context, contextSharing) {
  if (context) {
    return context;
  }

  if (contextSharing && typeof window !== 'undefined') {
    if (!window.SolidQueryClientContext) {
      window.SolidQueryClientContext = defaultContext;
    }

    return window.SolidQueryClientContext;
  }

  return defaultContext;
}



export const PersistQueryClientProvider = (props) => {
    const mergedProps = mergeProps({
        contextSharing: false,
    }, props);

    let persistQueryClientUnsubscribe

    onMount(() => {
      // same as in non-persisted provider
      //console.log('PersistQueryClientProvider.onMount: mount')
      mergedProps.client.mount()

      // persistence ...

      // note: different API
      // persistQueryClientSubscribe: props.queryClient
      // QueryClientProvider: props.client
      mergedProps.queryClient = mergedProps.client

      // subscribe
      //console.log('PersistQueryClientProvider.onMount: subscribe')
      persistQueryClientUnsubscribe =
      persistQueryClientSubscribe(mergedProps)

      // test: mount later
      //mergedProps.client.mount()
    });

    onCleanup(() => {
      // same as in non-persisted provider
      //console.log('PersistQueryClientProvider.onCleanup: unmount')
      mergedProps.client.unmount()

      // persistence ...

      //console.log('PersistQueryClientProvider.onCleanup: unsubscribe')
      // unsubscribe
      persistQueryClientUnsubscribe()
    })

    const QueryClientContext = getQueryClientContext(mergedProps.context, mergedProps.contextSharing);

    return (
      <QueryClientSharingContext.Provider value={!mergedProps.context && mergedProps.contextSharing}>
        <QueryClientContext.Provider value={mergedProps.client}>
          {mergedProps.children}
        </QueryClientContext.Provider>
      </QueryClientSharingContext.Provider>
    );
};
```

</details>

**Comment by lukesmurray** [comment]
Thank you for the comments @milahu. You figured out the naming logic before I had a chance to reply!

With regard to the missing apis.

- `useQueryErrorResetBoundary` was left out consciously because SolidJS has a [built-in error boundary](https://www.solidjs.com/docs/latest/api#errorboundary) while react does not.
- `useHydration` is left out because, in the interest of getting `solid-query` to the community as fast as possible, we released it without server-side rendering support. There are plans to add server-side rendering support, but I'm not sure if anyone is pursuing it at the moment. You can ask in the [discord](https://tlinz.com/discord). That's also probably the best place to talk to the maintainers about the current design and propose new features 😄.

About your error, you may want to use [unwrap](https://www.solidjs.com/docs/latest/api#unwrap) to get the store without a proxy.

About your proposal for the query key. I'm not quite sure if it would work. It depends on what `todo` is. But if `todo` is a signal, you're resolving the signal outside of `getQueryKey` so `getQueryKey` would not be reactive. 

```tsx
// only runs once
const queryKey = ["todos", todo()]

// runs multiple times but returns the non-reactive array every time
const getQueryKey = () => queryKey
```

In general, if you use [query key factories](https://tkdodo.eu/blog/effective-react-query-keys#use-query-key-factories) the fact that query keys are functions can be abstracted away fairly easily and that's probably the approach I would recommend.

**Pull request merged**
